### PR TITLE
Set requires_gradient to help autodiff to prune unneeded gradients

### LINKED ISSUE #54040: autodiff should NOT output grad_inputs for inputs which doesn't require_grad
## 🚀 Feature

autodiff in PyTorch JIT construct backward graph for DifferentiableGraph without considering requires_grad on inputs. This ended up with wasting computation to produce extra grad_inputs that is not used.

## Motivation

Extra computation is not idea. We observe the problem in mask used in attention used by NLP networks, where the extra grad_inputs computed ended up as reductions.

## Pitch

We should construct/optimize backward graph based on requires_grad flag on inputs. With @t-vi's PR: #49299, we are guarding requires_grad already and we should be able to use the mask to trim dummy grad_input that is not needed.

## Additional context

script as below:
```
import torch
@torch.jit.script
def t(input, bias):
  return torch.nn.functional.relu(input + bias)
input = torch.randn(2, 8, requires_grad=True)
bias = torch.randn(8, requires_grad=False)    # bias does NOT require grad
o = t(input, bias)
o.sum().backward()
o = t(input, bias)
o.sum().backward()
o = t(input, bias)
o.sum().backward()

print(t.graph_for(input, bias))
bwd_graph = list(list(t.get_debug_state().execution_plans.values())[0].code.grad_executor_states()[0].execution_plans.values())[0].graph
print(bwd_graph)
```

Forward graph:
```
graph(%input.1 : Float(2, 8, strides=[8, 1], requires_grad=1, device=cpu),
      %bias.1 : Float(8, strides=[1], requires_grad=0, device=cpu)):
  %2 : Float(8, strides=[1], requires_grad=0, device=cpu), %3 : Float(2, 8, strides=[8, 1], requires_grad=1, device=cpu), %4 : bool = prim::RequiresGradCheck[types=[Tensor(re
quires_grad=0), Tensor(requires_grad=1)]](%bias.1, %input.1)
  %5 : Tensor = prim::If(%4)
    block0():
      %6 : Tensor = prim::DifferentiableGraph_0(%2, %3)
      -> (%6)
    block1():
      %7 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %8 : (Tensor) = prim::CallFunction(%7, %bias.1, %input.1)
      %9 : Tensor = prim::TupleUnpack(%8)
      -> (%9)
  return (%5)
with prim::DifferentiableGraph_0 = graph(%0 : Tensor,
      %1 : Tensor):
  %2 : int = prim::Constant[value=1]()
  %5 : Tensor = aten::add(%1, %0, %2) # t.py:5:34
  %7 : int[] = aten::size(%1) # <string>:3:44
  %9 : int[] = aten::size(%5) # <string>:3:55
  %10 : int[]? = aten::_size_if_not_equal(%7, %9) # <string>:3:19
  %12 : int[] = aten::size(%0) # <string>:3:93
  %13 : int[]? = aten::_size_if_not_equal(%12, %9) # <string>:3:68
  %result.7 : Tensor = aten::relu(%5) # /opt/conda/lib/python3.6/site-packages/torch/nn/functional.py:1206:17
  return (%result.7, %10, %13, %result.7)
```

Where backward graph for DifferentiableGraph is as below. Note that it produces two grad_inputs, where `%grad_other.7` is produced for `bias` which doesn't require grad.
```
graph(%0 : Tensor,
      %1 : Tensor,
      %self_size.5 : int[]?,
      %other_size.5 : int[]?,
      %result.7 : Tensor):
  %71 : int = prim::Constant[value=0]() # <string>:98:47
  %191 : None = prim::Constant()
  %192 : bool = aten::__is__(%self_size.5, %191)
  %193 : bool = prim::AutogradAllNonZero(%1, %result.7)
  %194 : bool = prim::AutogradAllZero(%0)
  %195 : bool[] = prim::ListConstruct(%192, %193, %194)
  %196 : bool = aten::all(%195)
  %63 : (Tensor, Tensor) = prim::If(%196)
    block0():
      %85 : Tensor = aten::gt(%result.7, %71) # <string>:98:38
      %88 : Tensor = aten::type_as(%85, %result.7) # <string>:98:38
      %91 : Tensor = aten::mul(%1, %88) # <string>:98:23
      %grad_other.7 : Tensor = aten::_grad_sum_to_size(%91, %other_size.5) # <string>:17:30
      %126 : (Tensor, Tensor) = prim::TupleConstruct(%grad_other.7, %91)
      -> (%126)
    block1():
      %198 : Function = prim::Constant[name="fallback_function", fallback=1]()
      %199 : ((Tensor, Tensor)) = prim::CallFunction(%198, %0, %1, %self_size.5, %other_size.5, %result.7)
      %200 : (Tensor, Tensor) = prim::TupleUnpack(%199)
      -> (%200)
  return (%63)
```

cc @VitalyFedyunin @ngimel @gmagogsfm

### PR DESCRIPTION
**Pull request by Krovatkin**
Fixes https://github.com/pytorch/pytorch/issues/54040
`prim::RequiresGradCheck` guarantees that requires_grad properties
of input tensors will match the profiled, otherwise a fallback path
will be triggered. This allow us to prune off gradients in backward
graph for inputs that don't need gradients. We transfer requires_grad
properties from inputs to the `prim::DifferentiableGraph` onto inputs to the
differentiable graph. Autodiff will inspect these properties and prune
off gradients that aren't required

### CODE REVIEW

**Comment by eellison** [comment]
Can you add a comment or two to this code and PR?

**Review by eellison — APPROVED** [review]

**Review by jjsjann123 — APPROVED** [review]
LGTM
